src/csim/update_ops_control_single_target_single.cpp

```diff
@@ -251,3 +252,136 @@ void single_qubit_control_single_qubit_dense_matrix_gate_simd(
     }
 }
 #endif
+
+#ifdef _USE_MPI
+void single_qubit_control_single_qubit_dense_matrix_gate_mpi_OI(
+    UINT control_qubit_index, UINT control_value, CTYPE* t,
+    const CTYPE matrix[4], CTYPE* state, ITYPE dim, int flag,
+    UINT index_offset);
+void single_qubit_control_single_qubit_dense_matrix_gate_mpi_OO(
+    CTYPE* t, const CTYPE matrix[4], CTYPE* state, ITYPE dim, int flag);
+
+void single_qubit_control_single_qubit_dense_matrix_gate_mpi(
+    UINT control_qubit_index, UINT control_value, UINT target_qubit_index,
+    const CTYPE matrix[4], CTYPE* state, ITYPE dim, UINT inner_qc) {
+    MPIutil& m = MPIutil::get_inst();
+    const UINT rank = m.get_rank();
+    const UINT control_rank_bit = 1 << (control_qubit_index - inner_qc);
```

Comment:
Behavior of shifting by negative value is undefined.
Calculating those in if blocks is safer.

Reply:
Fixed and a little refactored at 8c0ad7c